odoo/models.py

```diff
@@ -2054,7 +2054,11 @@ def _read_group_process_groupby(self, gb, query):
             }
             if tz_convert:
                 qualified_field = "timezone('%s', timezone('UTC',%s))" % (self._context.get('tz', 'UTC'), qualified_field)
-            qualified_field = "date_trunc('%s', %s::timestamp)" % (gb_function or 'month', qualified_field)
+            week_start = babel.core.Locale.parse(get_lang(self.env).code).first_week_day
+            # 0 is monday, 6 is sunday
+            week_start = 0 if gb_function != 'week' else (7 - week_start) % 7
+            #If the week start is 1 (monday) we dont change anything, if it's 7 (sunday) the week should start 1 day earlier so 8 - 7 = 1 day early
+            qualified_field = "date_trunc('%s', %s::timestamp + INTERVAL '%s DAY') - INTERVAL '%s DAY'" % (gb_function or 'month', qualified_field, week_start, week_start)
```

Comment:
While it will get the week *boundaries* correct, it will not get the week *numbers* correct, because postgres makes its "week" computation based on the ISO week, which doesn't match traditional week numbering (to the extent that other calendars have a week numbering scheme).

I guess we can always tell people who don't use ISO weeks to fuck off, but if you start fixing the "issues" they report, they'll report new issues.

Reply:
I don't know either way, I'm just telling you what I see in the commit and the consequences that might yield. It's probably up to whoever the product owner is for this component to decide.

Reply:
the week number is gotten from using babel on the truncated date, and it depends on first day of the week ([babel/dates.py#L1509](https://github.com/python-babel/babel/blob/90739af21f7f26b126ffb52a136c624aba969acd/babel/dates.p#L1509)):

- if week start is on Monday => compute iso week/year correctly

- otherwise => first year day is week "1", on each new "first day of the week" the week number increases

So with this change, the week number would be correct according to babel.

Comparing it with calendar view in en_US, "Week 2" contains days from Sunday 2nd January 2022 to Saturday 8nd January 2022, in a read group we get:

- before the fix: "W2 2022" contains days from "Monday 3rd January 2022 to Sunday 9nd January 2022"
- after the fix: "W2 2022" contains the same day as the week shown in calendar

So I think the issue is more about if we want to keep iso boundary (with wrong week number when not in iso) or have locale boundary.

In another ticket (opw-2609234) there was a proposition (https://github.com/odoo/odoo/pull/74736) to do the contrary, have ISO week number but it was rejected by odo:

> This is not a bug, Odoo tries to display the week number in an ISO-like fashion, but which depends on the locale (country/language) of the user, in order to match *most* user expectations. We don't use strict ISO8601 semantics, in order to reduce frustrations and tickets, as many countries expect their local week numbers to be used and to match their first day of the week.
We keep the same convention in the various Calendar views that are available in our software too, for consistency, and this is where it matters even more.